src/tests/roc_ctl/test_task_queue.cpp

```diff
@@ -1290,5 +1290,53 @@ TEST(task_queue, reschedule_cancelled) {
     UNSIGNED_LONGS_EQUAL(1, tq.num_tasks());
 }
 
+TEST(task_queue, no_starvation) {
+    TestTaskQueue tq;
+    CHECK(tq.valid());
+
+    enum { NumTasks = 20 };
+
+    UNSIGNED_LONGS_EQUAL(0, tq.num_tasks());
+
+    TestHandler handler;
+    handler.expect_success(true);
+    handler.expect_n_calls(NumTasks);
+
+    TestTaskQueue::Task* tasks = new TestTaskQueue::Task[NumTasks];
+
+    tq.block();
+
+    const core::nanoseconds_t now = core::timestamp();
+    const core::nanoseconds_t WaitTime = core::Millisecond;
+
+    tq.schedule_at(tasks[0], now + WaitTime, &handler);
+    tq.set_nth_result(0, true);
+    for (size_t i = 1; i < NumTasks; i++) {
+        tq.set_nth_result(i, true);
+        tq.schedule(tasks[i], &handler);
+    }
+
+    // wait for sleeping task to sync
+    core::sleep_for(WaitTime);
+
+    TestTaskQueue::Task* temp = NULL;
+    for (size_t i = 0; i < NumTasks; i++) {
+        tq.unblock_one();
+        temp = handler.wait_called();
+        UNSIGNED_LONGS_EQUAL(i + 1, tq.num_tasks());
+    }
+
+    // check that the sleeping task wasn't last to get processed
+    CHECK(temp != &tasks[0]);
```

Comment:
Could we implement some stricter check here?

Currently we're checking that sleeeping task wont wait for NumTasks ready tasks. But we allow it to wait for NumTasks-1 ready tasks, for example.

Can we instead schedule a few sleeping and ready tasks and check that each sleeping task with expired deadline never waits more than one ready task?

Reply:
changed the unit test to check for alternating queue fetching as requested